ARTEMIS-1208 Do not use reconnect-atempts=-1 in tests

dudaerich opened the pull request:
If a connection to a broker is configured with parameter reconnect-atempts=-1 and something goes wrong with the broker, the test may hang indefinitely. It happens in cases where the session is created in main test thread.

michaelandrepearce commented:
How about other test areas, such as MultiThreadRandomReattachTestBase, ConsumerTest etc, doing a quick search it seems there's about 40 files with setReconnectAttempts(-1), i assume a large proportion of these are tests?

dudaerich commented:
You are right. Tomorrow I'll take a better look and update all occurrences of setReconnectAttempts in tests.

clebertsuconic commented:
@dudaerich  did you run the whole testsuite on this?

dudaerich commented:
@clebertsuconic Yes, I did. Currently there is a known issue with replicated HA and the number of failed tests in this area differs with each run. It is hard to say if this commit breaks some tests or not. I took a brief look on test results but I didn't identify any failure caused by some kind of error that connection was closed or something like that. All issues what I saw where related to either set up phase (backup wasn't synchronized with live in timeout) or clients didn't receive response on packet.

To be 100% sure we can increase reconnect-attempts or reconnect-delay. Currently reconnect-delay=50 and reconnect-attempts=200 what gives clients 10s to reconnect. Maybe we can increase the time to 30s. E.g. reconnect-delay=100 and reconnect-attempts=300 wdyt?

dudaerich commented:
I've changed the reconnect-attempts attribute in all tests. I've set it in a way to give clients 30s for reconnection.

Don't merge this please. I want to run the whole testsuite first.

clebertsuconic commented:
this increased the failure rate on a lot of FailoverTests, ReplicationTests.. etc.. failing because of this.



Are you sure this is an issue? ThreadLeak would find any threads still trying to reconnect.. that's actually why i developed that rule on ThreadLeaks.

clebertsuconic commented:
mainly failures are added by testCreateNewFactoryAfterFailover and testTimeoutOnFailoverConsumeBlocked...

a lot of the extending failover tests because of that...

Maybe it's easy to fix.. but I would like to see it solved before we merge.. otherwise it's hard to pinpoint the failure.


I want to understand also if this is really an issue.. as the ThreadLeak thingy would catch them.

dudaerich commented:
@clebertsuconic good catch! I'll take a look why testCreateNewFactoryAfterFailover fails.

I've attached thread dump from hanging test to https://issues.apache.org/jira/browse/ARTEMIS-1208.

Bellow are two threads which block the test. Main thread is waiting for failoverLock which is held by thread doing reconnection with reconnect-attempts=-1.
```
"main" Id=1 WAITING on java.util.concurrent.locks.ReentrantLock$NonfairSync@6e460863 owned by "Thread-4 (ActiveMQ-client-global-threads)" Id=5564 (in native)
	at sun.misc.Unsafe.park(Native Method)
	-  waiting on java.util.concurrent.locks.ReentrantLock$NonfairSync@6e460863
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:186)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:847)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireQueued(AbstractQueuedSynchronizer.java:881)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:1210)
	at java.util.concurrent.locks.ReentrantLock$NonfairSync.lock(ReentrantLock.java:220)
	at java.util.concurrent.locks.ReentrantLock.lock(ReentrantLock.java:296)
	at org.apache.activemq.artemis.core.client.impl.ClientSessionFactoryImpl.lockFailover(ClientSessionFactoryImpl.java:230)
	at org.apache.activemq.artemis.core.protocol.core.impl.ActiveMQClientProtocolManager.lockSessionCreation(ActiveMQClientProtocolManager.java:163)
	at org.apache.activemq.artemis.core.protocol.core.impl.ActiveMQClientProtocolManager.createSessionContext(ActiveMQClientProtocolManager.java:272)
	at org.apache.activemq.artemis.core.protocol.core.impl.ActiveMQClientProtocolManager.createSessionContext(ActiveMQClientProtocolManager.java:237)
	at org.apache.activemq.artemis.core.client.impl.ClientSessionFactoryImpl.createSessionChannel(ClientSessionFactoryImpl.java:1284)
	-  locked java.lang.Object@18ee7eb5
	at org.apache.activemq.artemis.core.client.impl.ClientSessionFactoryImpl.createSessionInternal(ClientSessionFactoryImpl.java:670)
	at org.apache.activemq.artemis.core.client.impl.ClientSessionFactoryImpl.createSession(ClientSessionFactoryImpl.java:323)
	at org.apache.activemq.artemis.tests.integration.cluster.failover.FailoverTest.createSession(FailoverTest.java:94)
	at org.apache.activemq.artemis.tests.integration.cluster.failover.FailoverTest.simpleFailover(FailoverTest.java:693)
	at org.apache.activemq.artemis.tests.integration.cluster.failover.FailoverTest.testFailBack(FailoverTest.java:534)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:95)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
	at java.lang.reflect.Method.invoke(Method.java:508)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:283)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:173)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:128)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:203)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:155)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)

"Thread-4 (ActiveMQ-client-global-threads)" Id=5564 TIMED_WAITING on java.util.concurrent.CountDownLatch$Sync@160a2aa1 (in native)
	at sun.misc.Unsafe.park(Native Method)
	-  waiting on java.util.concurrent.CountDownLatch$Sync@160a2aa1
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:226)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedNanos(AbstractQueuedSynchronizer.java:1048)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.tryAcquireSharedNanos(AbstractQueuedSynchronizer.java:1339)
	at java.util.concurrent.CountDownLatch.await(CountDownLatch.java:288)
	at org.apache.activemq.artemis.core.protocol.core.impl.ActiveMQClientProtocolManager.waitOnLatch(ActiveMQClientProtocolManager.java:136)
	at org.apache.activemq.artemis.core.client.impl.ClientSessionFactoryImpl.getConnectionWithRetry(ClientSessionFactoryImpl.java:819)
	at org.apache.activemq.artemis.core.client.impl.ClientSessionFactoryImpl.reconnectSessions(ClientSessionFactoryImpl.java:744)
	at org.apache.activemq.artemis.core.client.impl.ClientSessionFactoryImpl.failoverOrReconnect(ClientSessionFactoryImpl.java:614)
	at org.apache.activemq.artemis.core.client.impl.ClientSessionFactoryImpl.handleConnectionFailure(ClientSessionFactoryImpl.java:504)
	at org.apache.activemq.artemis.core.client.impl.ClientSessionFactoryImpl.access$500(ClientSessionFactoryImpl.java:72)
	at org.apache.activemq.artemis.core.client.impl.ClientSessionFactoryImpl$DelegatingFailureListener.connectionFailed(ClientSessionFactoryImpl.java:1165)
	at org.apache.activemq.artemis.spi.core.protocol.AbstractRemotingConnection.callFailureListeners(AbstractRemotingConnection.java:67)
	at org.apache.activemq.artemis.core.protocol.core.impl.RemotingConnectionImpl.fail(RemotingConnectionImpl.java:208)
	at org.apache.activemq.artemis.spi.core.protocol.AbstractRemotingConnection.fail(AbstractRemotingConnection.java:210)
	at org.apache.activemq.artemis.core.client.impl.ClientSessionFactoryImpl$CloseRunnable.run(ClientSessionFactoryImpl.java:996)
	at org.apache.activemq.artemis.utils.OrderedExecutorFactory$OrderedExecutor$ExecutorTask.run(OrderedExecutorFactory.java:101)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1153)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.lang.Thread.run(Thread.java:785)

	Number of locked synchronizers = 2
	- java.util.concurrent.locks.ReentrantLock$NonfairSync@6e460863
	- java.util.concurrent.ThreadPoolExecutor$Worker@e8c525c7
```

dudaerich commented:
I found out why testCreateNewFactoryAfterFailover failed. I changed the default value of reconnect-attempts from 0 to 300 and the test needed reconnect-attempts=0.

I've updated the PR. I threw away previous changes and I did it again. I didn't change locator created in FailoverTestBase. I changed only code which set up reconnect-attempts to -1.

I ran the modified test cases locally on my laptop and all tests passed.
